trie/stacktrie.go

```diff
@@ -129,6 +143,12 @@ const (
 )
 
 func (n *stNode) reset() *stNode {
+	if n.typ == hashedNode {
```

Comment:
What's the difference between HashedNode and Embedded tiny node?

For embedded node, we also allocate the buffer from the bPool and the buffer is owned by the node itself right?

Reply:
Well, the thing is, that there's only one place in the stacktrie where we convert a node into a `hashedNode`.
```golang
	st.typ = hashedNode
	st.key = st.key[:0]

	st.val = nil // Release reference to potentially externally held slice.
```
This is the only place. So we _know_ that once something has been turned into a hashedNode, the `val` is never externally held and thus it can be returned to the pool. The big problem we had earlier, is that we need to overwrite the `val` with a hash, but we are not allowed to mutate `val`. So this was a big cause of runaway allocs. 

But now we reclaim those values-which-are-hashes since we know that they're "ours".

Reply:
> For embedded node, we also allocate the buffer from the bPool and the buffer is owned by the node itself right?

To answer your question: yes. So from the perspective of slice-reuse, there is zero difference.

Reply:
Btw, in my follow-up PR to reduce allocs in derivesha:https://github.com/ethereum/go-ethereum/pull/30747 , I remove this trick again. 

In that PR, I _always_ copy the value as it enters the stacktrie. So we always own the `val` slice, and are free to reuse via pooling. 

Doing so is less hacky, we get rid of this special case "val is externally held _unless_ it's an hashedNode because then we own it". 

---

That makes it possible for the derivesha-method to reuse the input-buffer